### ISSUE TITLE
Error using Parsl with TorqueProvider

### ISSUE THREAD

**Issue by ricardoduarte** [Parsl/parsl#452]
Hi,

I have this simple code that I'm trying to run in a cluster with Torque installed:
```
@bash_app
def sleep():
    return 'sleep 30'

@click.command()
def test():
    config = Config(
        executors=[
            IPyParallelExecutor(
                provider=TorqueProvider(channel=LocalChannel()),
                label='local_ipp'
            )
        ]
    )
    parsl.load(config)
    sleep()
```
But, when I execute the code I'm getting the following error:
```
Traceback (most recent call last):
  File "/opt/python-3.6.6/lib/python3.6/site-packages/parsl/executors/ipp_controller.py", line 84, in start
    self.proc = subprocess.Popen(opts, stdout=stdout, stderr=stderr, preexec_fn=os.setsid)
  File "/opt/python-3.6.6/lib/python3.6/subprocess.py", line 709, in __init__
    restore_signals, start_new_session)
  File "/opt/python-3.6.6/lib/python3.6/subprocess.py", line 1344, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'ipcontroller': 'ipcontroller'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "utils.py", line 232, in <module>
    cli(prog_name='utils')
  File "/opt/python-3.6.6/lib/python3.6/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/opt/python-3.6.6/lib/python3.6/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/opt/python-3.6.6/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/python-3.6.6/lib/python3.6/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/opt/python-3.6.6/lib/python3.6/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "utils.py", line 221, in test
    parsl.load(config)
  File "/opt/python-3.6.6/lib/python3.6/site-packages/parsl/dataflow/dflow.py", line 898, in load
    cls._dfk = DataFlowKernel(config)
  File "/opt/python-3.6.6/lib/python3.6/site-packages/parsl/dataflow/dflow.py", line 121, in __init__
    executor.start()
  File "/opt/python-3.6.6/lib/python3.6/site-packages/parsl/executors/ipp.py", line 100, in start
    self.controller.start()
  File "/opt/python-3.6.6/lib/python3.6/site-packages/parsl/executors/ipp_controller.py", line 88, in start
    raise ControllerError(msg)
parsl.executors.errors.ControllerError: Controller init failed:Reason:Could not find ipcontroller. Please make sure that ipyparallel is installed and available in your env
```

If it helps to find the issue, I don't have root privilege in this cluster. Any idea why this is occurring is very appreciated.

**Comment by kylechard** [comment]
Hi Ricardo, 

Sorry you're seeing this error. You shouldn't need root access to run Parsl. What cluster are you running this on? Does the script work using a local IPP (i.e., not submitting to Torque)? E.g., if you use the following configuration: 

```
ipp_config = Config(
    executors=[
        IPyParallelExecutor(
            label="local_ipp",
            provider=LocalProvider(
                channel=LocalChannel()
            )
        )
    ]
)
```

**Comment by yadudoc** [comment]
In case you have trouble with the config Kyle posted above, you may have to explicitly set the IP address of the login node you are running parsl on. We have a some documentation here : http://parsl.readthedocs.io/en/latest/faq.html#workers-do-not-connect-back-to-parsl on setting the IP. Worst case, you could the IP address range used by the login nodes that is accessible from the compute nodes from the system admins.

**Comment by ricardoduarte** [comment]
I tested the config suggested by @kylechard and same error occurred.
We're running this on an old cluster, with CentOS 5 in it and Python 3.6.6 compiled by myself.

@yadudoc I tried setting the IP address too, but the error continues the same.

**Comment by benclifford** [comment]
Can you check that you have `ipcontroller` on your $PATH? That is the command that parsl is failing to start here.

eg in the same environment that you've run your python code, you should be able to do something like this:

```
$ ipcontroller
2018-08-09 05:43:39.751 [IPControllerApp] Hub listening on tcp://127.0.0.1:59131 for registration.
^C
```

**Comment by ricardoduarte** [comment]
You're right, `ipcontroller` wasn't on `$PATH`. Now, it worked with `LocalProvider`:
```
$ export PATH=$PATH:/opt/python-3.6.6/bin
$ /opt/python-3.6.6/bin/python3 utils.py test
Teste de implementacao da nova biblioteca parsl
DEBUG:parsl.executors.ipp:Starting IPyParallelExecutor with provider:
LocalProvider(
    channel=LocalChannel(
        envs={}, 
        script_dir='/mnt/STORAGE_GEOF/projetos_cipd_irps/ertm2d-bull/.scripts', 
        userhome='/mnt/STORAGE_GEOF/projetos_cipd_irps/ertm2d-bull'
    ), 
    init_blocks=4, 
    label='local', 
    launcher=<libsubmit.launchers.launchers.SingleNodeLauncher object at 0x2b057fd718d0>, 
    max_blocks=10, 
    min_blocks=0, 
    nodes_per_block=1, 
    parallelism=1, 
    script_dir='parsl_scripts', 
    tasks_per_node=1, 
    walltime='00:15:00'
)
DEBUG:parsl.executors.ipp:Launched block: 0:4230
DEBUG:parsl.executors.ipp:Launched block: 1:4231
DEBUG:parsl.executors.ipp:Launched block: 2:4232
DEBUG:parsl.executors.ipp:Launched block: 3:4233
DEBUG:parsl.executors.ipp:Starting executor
DEBUG:parsl.dataflow.strategy:Scaling strategy: simple
INFO:parsl.dataflow.dflow:Task 0 submitted for App remote_side_bash_executor, waiting on tasks []
INFO:parsl.dataflow.dflow:Task 0 launched on executor local_ipp
DEBUG:parsl.dataflow.dflow:Task 0 launched with AppFuture: <AppFuture at 0x2aaaaaaf0f98 state=pending>
INFO:parsl.dataflow.dflow:DFK cleanup initiated
INFO:parsl.dataflow.dflow:Summary of tasks in DFK:
INFO:parsl.dataflow.dflow:Tasks in state States.running: 0
INFO:parsl.dataflow.dflow:End of summary
INFO:parsl.dataflow.dflow:Terminating flow_control and strategy threads
DEBUG:parsl.executors.ipp:IPP:Shutdown sequence: Attempting controller kill
DEBUG:parsl.executors.ipp_controller:Controller exited with 0
DEBUG:parsl.executors.ipp:Done with executor shutdown
DEBUG:parsl.data_provider.data_manager:Done with executor shutdown
INFO:parsl.dataflow.dflow:DFK cleanup complete
```
But I'm getting a new error when I try with `TorqueProvider`:
```
[lslf@s6506ggf130 ertm2d-bull]$ /opt/python-3.6.6/bin/python3 utils.py test
Teste de implementacao da nova biblioteca parsl
DEBUG:parsl.executors.ipp:Starting IPyParallelExecutor with provider:
TorqueProvider(
    LocalChannel(
        envs={}, 
        script_dir='/mnt/STORAGE_GEOF/projetos_cipd_irps/ertm2d-bull/.scripts', 
        userhome='/mnt/STORAGE_GEOF/projetos_cipd_irps/ertm2d-bull'
    ),
    account=None, 
    init_blocks=1, 
    label='torque', 
    launcher=<libsubmit.launchers.launchers.AprunLauncher object at 0x2afedc790b00>, 
    max_blocks=100, 
    min_blocks=0, 
    nodes_per_block=1, 
    overrides='', 
    parallelism=1, 
    queue=None, 
    script_dir='parsl_scripts', 
    tasks_per_node=1, 
    walltime=21
)
Caught exception : Command 'qsub  /mnt/STORAGE_GEOF/projetos_cipd_irps/ertm2d-bull/.scripts/parsl.parsl.auto.1533823164.113927.submit' timed out after 3 seconds
WARNING:libsubmit.channels.local.local:Execution of command [qsub  /mnt/STORAGE_GEOF/projetos_cipd_irps/ertm2d-bull/.scripts/parsl.parsl.auto.1533823164.113927.submit] failed due to 
 Command 'qsub  /mnt/STORAGE_GEOF/projetos_cipd_irps/ertm2d-bull/.scripts/parsl.parsl.auto.1533823164.113927.submit' timed out after 3 seconds 
ERROR:parsl.executors.ipp:Scaling out failed: 'NoneType' object has no attribute 'strip'
Traceback (most recent call last):
  File "utils.py", line 234, in <module>
    cli(prog_name='utils')
  File "/opt/python-3.6.6/lib/python3.6/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/opt/python-3.6.6/lib/python3.6/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/opt/python-3.6.6/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/python-3.6.6/lib/python3.6/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/opt/python-3.6.6/lib/python3.6/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "utils.py", line 223, in test
    parsl.load(config)
  File "/opt/python-3.6.6/lib/python3.6/site-packages/parsl/dataflow/dflow.py", line 898, in load
    cls._dfk = DataFlowKernel(config)
  File "/opt/python-3.6.6/lib/python3.6/site-packages/parsl/dataflow/dflow.py", line 121, in __init__
    executor.start()
  File "/opt/python-3.6.6/lib/python3.6/site-packages/parsl/executors/ipp.py", line 135, in start
    raise e
  File "/opt/python-3.6.6/lib/python3.6/site-packages/parsl/executors/ipp.py", line 126, in start
    engine = self.provider.submit(self.launch_cmd, 1)
  File "/opt/python-3.6.6/lib/python3.6/site-packages/libsubmit/providers/torque/torque.py", line 220, in submit
    logger.error("Retcode:%s STDOUT:%s STDERR:%s", retcode, stdout.strip(), stderr.strip())
AttributeError: 'NoneType' object has no attribute 'strip'
```
One doubt: Is it mandatory to have `ipcontroller` on $PATH, or can I somehow set the ipcontroller PATH in the config?

**Comment by benclifford** [comment]
That's a messy stack track which appears when submission of a job to torque using the `qsub` command line fails or times-out (see it happening with slurm, not torque here: https://github.com/Parsl/parsl/issues/419 ).

The command it is trying to run in your case is:
 ```
qsub  /mnt/STORAGE_GEOF/projetos_cipd_irps/ertm2d-bull/.scripts/parsl.parsl.auto.1533823164.113927.submit
```

which you might be able to try by hand.

The command has to complete within 3 seconds, because there is a hard-coded 3 second timeout in the parsl/lisubmit torque code.

If `qsub` takes longer than this, and you can edit the libsubmit code, try editing: `libsubmit/providers/torque/torque.py`

There are 3 instances of the numeric literal `3` in that file - all of which you might make much bigger - for example, to 60 (seconds). 

I've experienced similar timeouts when using slurm on a large system - because the slurm commands take a while on such a system.

your other question:
```One doubt: Is it mandatory to have ipcontroller on $PATH, or can I somehow set the ipcontroller PATH in the config?```

It's pretty mandatory.

However, I think you can set the PATH within from your python code, before initialising parsl, if that is how you would prefer to do things - for example, `import os`  and then modify  `os.environ['PATH']` near the start of your code.

**Comment by annawoodard** [comment]
Looking at the trace, it looks like the path to the file you'd need to change to increase the timeout is

```
/opt/python-3.6.6/lib/python3.6/site-packages/libsubmit/providers/torque/torque.py
```

Let us know if that fixes your issue, and we'll prepare a patch to make the timeout configurable.

**Comment by ricardoduarte** [comment]
Thank you for your excellent help. I tried `qsub` by hand and the timeout is not the real problem. One or more nodes of the cluster is offline, and that is the root cause of the error.

I am getting in contact with someone inside our company to help me with that, and once it is solved, I can test the implementation again.

Congrats for this library and the excellent job that you guys are doing maintaining it.

**Comment by ricardoduarte** [comment]
Hi guys,
The node is alive again, but now I have a new error:
```
$ python3 utils.py test
Teste de implementacao da nova biblioteca parsl
DEBUG:parsl.executors.ipp:Starting IPyParallelExecutor with provider:
TorqueProvider(
    LocalChannel(
        envs={}, 
        script_dir='/mnt/STORAGE_GEOF/projetos_cipd_irps/ertm2d-bull/.scripts', 
        userhome='/mnt/STORAGE_GEOF/projetos_cipd_irps/ertm2d-bull'
    ),
    account=None, 
    init_blocks=1, 
    label='torque', 
    launcher=<libsubmit.launchers.launchers.AprunLauncher object at 0x2b3707c40be0>, 
    max_blocks=100, 
    min_blocks=0, 
    nodes_per_block=1, 
    overrides='', 
    parallelism=1, 
    queue=None, 
    script_dir='parsl_scripts', 
    tasks_per_node=1, 
    walltime=21
)
DEBUG:parsl.executors.ipp:Launched block: 0:154450.s6506ggf132.cenpes.petrobras.com.br
DEBUG:parsl.executors.ipp:Starting executor
DEBUG:parsl.dataflow.strategy:Scaling strategy: simple
INFO:parsl.dataflow.dflow:Task 0 submitted for App remote_side_bash_executor, waiting on tasks []
INFO:parsl.dataflow.dflow:Task 0 launched on executor local_ipp
DEBUG:parsl.dataflow.dflow:Task 0 launched with AppFuture: <AppFuture at 0x2aaaaaac4048 state=pending>
INFO:parsl.dataflow.dflow:DFK cleanup initiated
INFO:parsl.dataflow.dflow:Summary of tasks in DFK:
INFO:parsl.dataflow.dflow:Tasks in state States.running: 0
INFO:parsl.dataflow.dflow:End of summary
INFO:parsl.dataflow.dflow:Terminating flow_control and strategy threads
Error in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/opt/python-3.6.6/lib/python3.6/site-packages/parsl/dataflow/dflow.py", line 665, in atexit_cleanup
    self.cleanup()
  File "/opt/python-3.6.6/lib/python3.6/site-packages/parsl/dataflow/dflow.py", line 707, in cleanup
    executor.scale_in(len(job_ids))
  File "/opt/python-3.6.6/lib/python3.6/site-packages/parsl/executors/ipp.py", line 253, in scale_in
    status = dict(zip(self.engines, self.provider.status(self.engines)))
  File "/opt/python-3.6.6/lib/python3.6/site-packages/libsubmit/providers/cluster_provider.py", line 176, in status
    self._status()
  File "/opt/python-3.6.6/lib/python3.6/site-packages/libsubmit/providers/torque/torque.py", line 130, in _status
    self.resources[job_id]['status'] = status
KeyError: '154450.s6506ggf132'
```
The job was created correctly, when I issued a `qstat` command right after the error, I got this output:
```
$ qstat
Job id                    Name             User            Time Use S Queue
------------------------- ---------------- --------------- -------- - -----
154450.s6506ggf132         ...02984.0736158 lslf            00:00:00 C batch 
```

**Comment by annawoodard** [comment]
Could this be related to #313?

I see that the DFK is getting cleaned up while a task is still running. If you have a very simple script e.g.

```
@python_app
def foobar():
    return True

foobar()
```

where you never call result, the DFK might clean up without waiting for the task. Instead, you'd need

```
@python_app
def foobar():
    return True

foobar().result()
```

to force waiting for the result (we will fix this in the above-linked issue).

**Comment by ricardoduarte** [comment]
I tried now call `sleep().result()` and it gave me the same error. The difference is that it doesn't return the command line control to me. I had to finalize it with <CTRL> + <C>.
I am suspecting that while it cannot find the task status, it will never end the task.

**Comment by annawoodard** [comment]
Note we separately need to deal with the `KeyError: '154450.s6506ggf132'` which looks to just be a matter of trying to check the status after we've already killed the engine.

**Comment by annawoodard** [comment]
Can you provide the script you're trying to run?

**Comment by ricardoduarte** [comment]
Here is the code:
```
@bash_app
def sleep():
    return 'sleep 30'

@click.command()
def test():
    """
    Comando para testar a implementacao do parsl
    """
    click.echo('Teste de implementacao da nova biblioteca parsl')
    config = Config(
        executors=[
            IPyParallelExecutor(
                # max_threads=2,
                provider=TorqueProvider(channel=LocalChannel()),
                label='local_ipp'
                # controller=Controller(public_ip='10.30.120.130')
            )
        ]
    )
    parsl.load(config)
    sleep().result()
```

**Comment by annawoodard** [comment]
Can you verify things are working with a local config? That would help us determine if it's a Torque issue, or something else. So I'd recommend:

```python
@click.command()
def test():
    """
    Comando para testar a implementacao do parsl
    """
    click.echo('Teste de implementacao da nova biblioteca parsl')
    parsl.set_stream_logger()
    from parsl.configs.local_ipp import config
    parsl.load(config)

    sleep().result()
```

Note that I think the `KeyError` is a separate issue. @yadudoc has access to a Torque cluster to test this, but he's away at the moment. I've applied for an account, so as soon as I'm approved, I can test a patch to fix the `KeyError`.

**Comment by ricardoduarte** [comment]
@annawoodard that implementation above ran without errors.

**Comment by annawoodard** [comment]
Hi Ricardo, sorry about the delay. I am not able to reproduce your error, because I don't have access to a Torque system. Our dev that has access to a Torque system for testing is out of town for the month. I am trying to get access, but it's been delayed. I'll report back here as soon as I get it.

**Comment by yadudoc** [comment]
@ricardoduarte  Hey, I have access to a cluster that I can test on. Since the last communication has anything changed ?

**Comment by ricardoduarte** [comment]
Hi @yadudoc, nothing changed. I would appreciate your help.

**Comment by yadudoc** [comment]
Hi @ricardoduarte, I'm sorry I dropped the ball on this. If you are still waiting on this one, I highly recommend that we could find out what is happening with a bit of one-on-one time over slack. Here's a link to our slack group:  [Link](https://join.slack.com/t/parsl-project/shared_invite/enQtMzg2MDAwNjg2ODY1LTk0ZmYyZWE2NDMwYzVjZThmNTUxOWE0MzNkN2JmYjMyY2QzYzg0YTM3MDEzYjc2ZjcxZGZhMGQ1MzBmOWRiOTM)

**Comment by ricardoduarte** [comment]
No problem @yadudoc. I have important work problems to solve this week and next week I'm leaving on vacation. Can we see this when I come back?

**Comment by yadudoc** [comment]
@ricardoduarte Of course. Ping us when you are ready to take a look at this issue again.

**Comment by ricardoduarte** [comment]
Hi all,
At this moment, we decided not to use Parsl in our project. We hope we can use it again in the future. Closing issue now.
Thank you for the support.
